python/ray/dashboard/modules/state/state_head.py

```diff
@@ -237,54 +228,33 @@ async def get_actor_fn(actor_id: ActorID) -> Optional[ActorTableData]:
-                    await response.write(logs)
-            except asyncio.CancelledError:
-                # This happens when the client side closes the connection.
-                # Force close the connection and do no-op.
-                response.force_close()
-                raise
-            except Exception as e:
-                logger.exception("Error while streaming logs")
-                await response.write(f"[get_logs] Fetch log error: {e}".encode())
-
-        async def formatter_leading_1(response, async_gen: AsyncIterable[bytes]):
-            # NOTE: The first byte indicates the success / failure of individual
-            # stream. If the first byte is b"1", it means the stream was successful.
-            # If it is b"0", it means it is failed.
-            try:
-                async for logs in async_gen:
-                    logs_to_stream = bytearray(b"1")
-                    logs_to_stream.extend(logs)
-                    await response.write(bytes(logs_to_stream))
-            except asyncio.CancelledError:
-                # This happens when the client side closes the connection.
-                # Fofce close the connection and do no-op.
-                response.force_close()
-                raise
-            except Exception as e:
-                logger.exception("Error while streaming logs")
-                error_msg = bytearray(b"0")
-                error_msg.extend(
-                    f"Closing HTTP stream due to internal server error.\n{e}".encode()
-                )
-                await response.write(bytes(error_msg))
-
         response = aiohttp.web.StreamResponse()
         response.content_type = "text/plain"
-        await response.prepare(req)
 
         logs_gen = self._log_api.stream_logs(options, get_actor_fn)
-        if output_format == "text":
-            await formatter_text(response, logs_gen)
-        elif output_format == "leading_1":
-            await formatter_leading_1(response, logs_gen)
-        else:
-            raise ValueError(
-                f"Unsupported format: {output_format}, use 'text' or " "'leading_1'"
-            )
+        # Handle the first chunk separately and returns 500 if an error occurs.
+        try:
+            first_chunk = await logs_gen.__anext__()
+            await response.prepare(req)
+            await response.write(first_chunk)
+        except StopAsyncIteration:
+            pass
+        except asyncio.CancelledError:
+            # This happens when the client side closes the connection.
+            # Force close the connection and do no-op.
+            response.force_close()
+            raise
+        except Exception as e:
+            logger.exception("Error while streaming logs")
+            raise aiohttp.web.HTTPInternalServerError(text=str(e))
+
+        try:
+            async for logs in logs_gen:
+                await response.write(logs)
+        except Exception:
+            response.force_close()
```

Comment:
why the exception handling for the first chunk is different from other chunks?

Reply:
Should we do `await response.write(f"[get_logs] Fetch log error: {e}".encode())`?

Reply:
https://github.com/ray-project/ray/blob/b2a986b3e962915f89ccac8c788f5c3723cb594d/python/ray/dashboard/modules/log/log_manager.py#L81-L127

Before the first chunk is generated, it does things like resolving the filename, creating a gRPC stub to connect to the log service, and initializing metadata. After the first chunk, the stream is guaranteed to exist. Therefore, I handle the first chunk differently from the subsequent ones.

Reply:
> Should we do await response.write(f"[get_logs] Fetch log error: {e}".encode())?

The error may also be a response write error. In that case, writing to the response again would raise another error, so I didn’t write to the response again.

Reply:
After this PR, how can client tell that second chunk has error?

Reply:
I don't think there's a reliable way for the client to distinguish stream errors after the first chunk is generated, because the chunks may be split. In the extreme case, the client may receive only a single byte, so it's hard to tell whether an error occurred or not.